src/lib.rs

```diff
@@ -122,38 +131,81 @@ mod platform {
         Ok(task)
     }
 
-    impl CopyAddress for Process {
-        fn copy_address(&self, addr: usize, buf: &mut [u8]) -> io::Result<()> {
-            let task = task_for_pid(self.pid);
-            if task.is_err() { return task.map(|_| ()) }
+    const VM_PAGE_SIZE: vm_size_t = 4096;
+    const VM_PAGE_MASK: vm_size_t = 4095;
 
-            let page_addr      = (addr as i64 & (-4096)) as mach_vm_address_t;
-	        let last_page_addr = ((addr as i64 + buf.len() as i64 + 4095) & (-4096)) as mach_vm_address_t;
-            let page_size      = last_page_addr as usize - page_addr as usize;
+    /// Adapted from:
+    ///   https://github.com/opensource-apple/xnu/blob/27ffc00/libsyscall/mach/mach/vm_page_size.h#L52-L56
+    fn mach_vm_trunc_page(x: mach_vm_offset_t) -> mach_vm_offset_t {
+        // NOTE: `!` is a bitwise NOT for integers in Rust (like `~` in C)
+        (x & !VM_PAGE_MASK)
+    }
+
+    fn mach_vm_round_page(x: mach_vm_offset_t) -> mach_vm_offset_t {
+        ((x + VM_PAGE_MASK) & !VM_PAGE_MASK)
+    }
+
+    /// Read multiple pages from a task/process's memory. Address arguments
+    /// *must* be page-aligned or else bad things will happen. For
+    /// page-alignment try `mach_vm_trunc_page` and `mach_vm_round_page`.
+    fn read_aligned(task: mach_port_name_t, start_addr: mach_vm_address_t, end_addr: mach_vm_address_t) -> io::Result<Vec<u8>> {
+        debug!("read_aligned: start_addr: {:x}", start_addr);
+        debug!("read_aligned: end_addr: {:x}", end_addr);
+
+        let num_pages = (end_addr - start_addr) / VM_PAGE_SIZE;
+
+        let mut buf: Vec<u8> = vec![];
+
+        for i in 0..num_pages {
+            let page_addr = start_addr + (i * VM_PAGE_SIZE);
 
             let read_ptr: *mut u8 = ptr::null_mut();
             let mut read_len: mach_msg_type_number_t = 0;
 
             let result = unsafe {
-                vm_read(task.unwrap(), page_addr as u64, page_size as vm_size_t, &read_ptr, &mut read_len)
+                vm_read(task, page_addr as u64, VM_PAGE_SIZE as vm_size_t, &read_ptr, &mut read_len)
             };
 
             if result != KERN_SUCCESS {
                 // panic!("({}) {:?}", result, io::Error::last_os_error());
                 return Err(io::Error::last_os_error())
             }
 
-            if read_len != page_size as u32 {
-                panic!("Mismatched read sizes for `vm_read` (expected {}, got {})", page_size, read_len)
+            if read_len != VM_PAGE_SIZE as u32 {
+                panic!("Mismatched read sizes for `vm_read` (expected {}, got {})", VM_PAGE_SIZE, read_len)
             }
 
             let read_buf = unsafe { slice::from_raw_parts(read_ptr, read_len as usize) };
 
-            let offset = addr - page_addr as usize;
-            let len = buf.len();
-            buf.copy_from_slice(&read_buf[offset..(offset + len)]);
+            buf.extend_from_slice(read_buf);
```

Comment:
does this still result in "polluting" vm space as new mappings are made by `vm_read`? if so, please add a FIXME mentioning it, so we know it still needs to be sorted out.